### DISCUSSION TITLE
Using AWS IAM for authentication

### DISCUSSION THREAD

**Discussion by jonapgartwohat** [prisma/prisma#4147] [Q&A] [answered]
I have been using Prisma on a prototype for a project and while I absolutely love it, I have come across a snag. 

My organization uses [IAM for authentication on Amazon RDS](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.html) and it appears Prisma will not work well with this method. I could of course generate a password token for RDS prior to starting my app, but the tokens expire every 15 minutes and I can't think of a good away of handling the reconnect. 

Do you have plans to support IAM or other arbitrary async authentication mechanisms? Is there a way to achieve this now and I just haven't found it yet? 

Thanks,
Jon

**Comment by ryands17** [comment]
Hey @jonapgartwohat 👋 
It should be possible to use IAM Authentication with Prisma (I have personally not tried it), but you would need to manually disconnect Prisma and connect again after the token expires.

You could add your DB URL in this manner as described [here](https://github.com/prisma/prisma/issues/2443#issuecomment-630679118)

```ts
import { PrismaClient } from '@prisma/client'

const client = new PrismaClient({ datasources: { mydb: { url: 'postgres://localhost/db1' }} })
```

This will override the URL and you would be able to pass the token as the password. In this case, you would need a mechanism to reconnect if the token has expired which you would need to manage yourself.

Also IAM DB Auth requires SSL so you would need to connect Prisma to your DB with SSL first.

Let me know if this helps :)

**Reply by ryands17** [reply]
Hey @jonapgartwohat 👋 
Did the above setup work? If so, could you share a sample code that you have used with IAM auth?

**Reply by jonapgartwohat** [reply]
Yes it did, although still building out some tests for error handling/disconnects etc. The `expiry` promise sent here is for the factory function to know when to pre-emptively disconnect/reconnect. DB_TOKEN_EXPIRY_SECONDS should be set to something less than the actual token expiry time for RDS (which I believe is 15 minutes).



```typescript
import { PrismaClientOptions } from '@prisma/client'
import { RDS } from 'aws-sdk'
//...
async function generate(): Promise<{ prismaOptions: PrismaClientOptions; expiry?: Promise<void> }> {
 
    let password: string
    let expiry: Promise<void> | undefined
    if (!DB_PASSWORD) {
      password = await generateRdsPassword({
        credentials: {
          accessKeyId: AWS_ACCESS_KEY,
          secretAccessKey: AWS_SECRET_KEY,
        },
        region: DB_REGION,
        hostname: DB_HOST,
        username: DB_USER,
        port: DB_PORT,
      })
      expiry = new Promise(res => setTimeout(res, parseInt(DB_TOKEN_EXPIRY_SECONDS, 10) * 1000))
    } else {
      password = DB_PASSWORD
    }

    const url = new URL(`postgresql://${DB_HOST}:${DB_PORT}/${DB_DATABASE}`)
    url.username = DB_USER
    url.password = password
    url.search = new URLSearchParams({
      schema: DB_SCHEMA,
      connection_timeout: DB_CONNECTION_TIMEOUT,
      socket_timeout: DB_QUERY_TIMEOUT,
      connection_limit: DB_MAX_CONNECTIONS,
      ...(RDS_ROOT_CERT_PATH
        ? {
            sslmode: 'verify-full',
            sslrootcert: RDS_ROOT_CERT_PATH,
          }
        : undefined),
    }).toString()

    return {
      prismaOptions: {
        datasources: {
          db: {
            url: url.toString(),
          },
        },
      },
      expiry,
    }
  }

//... 
function generateRdsPassword(options: RDS.Signer.SignerOptions): Promise<string> {
  //TODO cache signer?
  const signer = new RDS.Signer(options)
  return new Promise((res, rej) => {
    signer.getAuthToken(options, (err, token) => {
      err ? rej(err) : res(token)
    })
  })
}
```

**Reply by kevinchevalier** [reply]
You don't really need to disconnect though, right?  Only the password expires, not the open connections.

Is there a way to use a different password only when a new connection is needed for the pool?

**Comment by kevinchevalier** [comment]
We decided to use Prisma with IAM by creating a new Prisma client for every HTTP request that comes through the server.

This is a huge waste, and is only possible because we have so few active users.  

It's a shame that this can't be accomplished with Prisma, when something like Knex makes it so easy.  Wish Knex, you can set an `expirationChecker` that checks to see if the token is expired only when creating a new connection.  If it is, the token can be regenerated.

**Reply by ryands17** [reply]
@kevinchevalier 👋 

You would have to disconnect as creating a new `PrismaClient` for every HTTP request is not the recommended way. In this case, it would be much easier to disconnect and update the client with the new token that you get after expiration.

I would also suggest creating a feature request [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) so that we can look into this as it seems like an important request.

**Reply by kevinchevalier** [reply]
Thanks for the info.  Can you describe how the client can be updated after
it is created?

**Reply by ryands17** [reply]
just assigning new `PrismaClient` instance to your database variable after disconnecting should do the trick.

**Reply by kevinchevalier** [reply]
Thank you for helping me with this question, I really appreciate it.  

The part that I don't understand though is that I don't want to disconnect existing connections in the pool.  They are still valid.  Only the password expires.

There also might be other requests in the express pipeline that are actively using the same PrismaClient concurrently, so I wouldn't want to force a disconnect.

Please let me know if I'm misunderstanding something about how Prisma works, under the hood.

**Reply by ryands17** [reply]
Currently a `disconnect` is required because there isn't any mechanism that will update the same connection with the new token. We will be looking into the request you opened and think of a better way of managing this.

> There also might be other requests in the express pipeline that are actively using the same PrismaClient concurrently, so I wouldn't want to force a disconnect.

I don't think this will matter as these requests would fail anyway due to an expired token.

**Reply by kevinchevalier** [reply]
> I don't think this will matter as these requests would fail anyway due to an expired token.

If there are existing active connections in the connection pool, the requests will succeed.  The way that RDS with IAM works is that only the token expires.  The connections, once created, do not expire when the token expires.

Token expiration only affects the ability to create new connections to the database.

**Reply by kevinchevalier** [reply]
> Currently a disconnect is required because there isn't any mechanism that will update the same connection with the new token. We will be looking into the request you opened and think of a better way of managing this.

Great, thanks for the information.

**Reply by ThrowsException** [reply]
Whats the current state of this strategy? using `@prisma/client": "^3.11.1` and I'm getting this error even when calling disconnect 

`warn(prisma-client) There are already 10 instances of Prisma Client actively running.`

I'm also trying to use IAM auth using code that tries to create the initial prisma client then on setInterva disconnect, get a new token and reconnect but it looks like it just piles up new Prisma clients rather than reusing the existing one with a new connection string. 


```js
let prisma;
const generateClient = async (server, options) => {
  const signer = new RDS.Signer({
    // configure options
    region: 'us-east-1',
    username: 'iam_user',
    hostname: 'db',
    port: 5432,
  });
  const token = signer.getAuthToken();
  prisma = new PrismaClient({
    datasources: {
      db: {
        url: `postgresql://iam_user:${encodeURIComponent(
          token
        )}@db:5432/test?schema=public`,
      },
    },
    log: ['query', 'info', 'warn', 'error'],
  });

  setInterval(async () => {
    await prisma.$disconnect();
    const token = signer.getAuthToken();
    prisma = new PrismaClient({
      datasources: {
        db: {
          url: `postgresql://iam_user:${encodeURIComponent(
            token
          )}@db/test?schema=public`,
        },
      },
      log: ['query', 'info', 'warn', 'error'],
    });
  }, 2000);

  await prisma.$connect();
};
generateClient();

module.exports = prisma;
```

**Reply by nurul3101** [reply]
@ThrowsException 👋 

The solution in this case is to instantiate a single instance PrismaClient and save it on the [global](https://nodejs.org/api/globals.html#globals_global) object and then keeping a check to only instantiate PrismaClient if it's not on the global object otherwise using the same instance again if already present to prevent instantiating extra PrismaClient instances.

Here's the [Help Article](https://www.prisma.io/docs/support/help-articles/nextjs-prisma-client-dev-practices) which discusses the solution to this error.

**Reply by ThrowsException** [reply]
@nurul3101 Thanks for the suggestion but this still doesn't appear to be a solution for IAM. Even using global object i'm still getting the above error message. It seems no matter what calling `new PrismaClient` in the setInterval when I need to refresh the iam auth token results in more connections being opened even if I call $disconnect. Not exactly sure what I'm doing wrong there.

**Comment by kevinchevalier** [comment, marked as answer]
Added issue #7869 with a proposed solution.

**Reply by janpio** [reply]
And here is a RDS specific issue: https://github.com/prisma/prisma/issues/6587

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
